pkg/mimirtool/client/backfill.go

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/fs"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+)
+
+func (c *MimirClient) Backfill(ctx context.Context, source string, logger log.Logger) error {
+	// Scan blocks in source directory
+	es, err := os.ReadDir(source)
+	if err != nil {
+		return errors.Wrapf(err, "failed to read directory %q", source)
+	}
+	for _, e := range es {
+		if err := c.backfillBlock(ctx, filepath.Join(source, e.Name()), logger); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *MimirClient) backfillBlock(ctx context.Context, dpath string, logger log.Logger) error {
+	metaPath := filepath.Join(dpath, "meta.json")
+	f, err := os.Open(metaPath)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %q", metaPath)
+	}
+
+	var blockMeta metadata.Meta
+	if err := json.NewDecoder(f).Decode(&blockMeta); err != nil {
+		return errors.Wrapf(err, "failed to decode %q", metaPath)
+	}
+	if len(blockMeta.Thanos.Files) == 0 {
+		return fmt.Errorf("no files listed in meta.json")
+	}
+
+	blockID := filepath.Base(dpath)
+
+	level.Info(logger).Log("msg", "Making request to start block backfill", "user", c.id, "block_id", blockID)
+
+	buf := bytes.NewBuffer(nil)
+	if err := json.NewEncoder(buf).Encode(blockMeta); err != nil {
+		return errors.Wrap(err, "failed to JSON encode payload")
+	}
+	res, err := c.doRequest(fmt.Sprintf("/api/v1/upload/block/%s", blockID), http.MethodPost, buf,
+		int64(buf.Len()))
+	if err != nil {
+		return errors.Wrap(err, "request to start backfill failed")
+	}
+	defer res.Body.Close()
+	if res.StatusCode/100 != 2 {
+		return fmt.Errorf("request to start backfill failed, status code %d", res.StatusCode)
```

Comment:
Let's also include status too, in case server sends something useful there (at all places)

```suggestion
		return fmt.Errorf("request to start backfill failed, with HTTP status %d %s", res.StatusCode, res.Status)
```

Reply:
Done, PTAL.